records/generator.go

```diff
@@ -244,106 +183,105 @@ func hostToIP4(hostname string) (string, bool) {
 	return ip.String(), true
 }
 
-func (t *Task) containerIP() string {
-	const containerIPTaskStatusLabel = "Docker.NetworkSettings.IPAddress"
-
-	// find TASK_RUNNING statuses
-	var latestContainerIP string
-	var latestTimestamp float64
-	for _, status := range t.Statuses {
-		if status.State != "TASK_RUNNING" {
-			continue
-		}
-
-		// find the latest docker-inspect label
-		for _, label := range status.Labels {
-			if label.Key == containerIPTaskStatusLabel && status.Timestamp > latestTimestamp {
-				latestContainerIP = label.Value
-				latestTimestamp = status.Timestamp
-				break
-			}
-		}
-	}
-
-	return latestContainerIP
-}
-
 // InsertState transforms a StateJSON into RecordGenerator RRs
-func (rg *RecordGenerator) InsertState(sj StateJSON, domain string, ns string,
-	listener string, masters []string, spec labels.Func) error {
+func (rg *RecordGenerator) InsertState(sj state.State, domain string,
+	ns string, listener string, masters []string, spec labels.Func) error {
 
-	// creates a map with slave IP addresses (IPv4)
 	rg.SlaveIPs = make(map[string]string)
 	rg.SRVs = make(rrs)
 	rg.As = make(rrs)
 
-	for _, slave := range sj.Slaves {
-		address, ok := hostToIP4(slave.PID.Host)
-		if ok {
-			a := "slave." + domain + "."
-			rg.insertRR(a, address, "A")
-			srv := net.JoinHostPort(a, slave.PID.Port)
-			rg.insertRR("_slave._tcp."+domain+".", srv, "SRV")
-		} else {
-			logging.VeryVerbose.Printf("string '%q' for slave with id %q is not a valid IP address", address, slave.ID)
-			address = labels.DomainFrag(address, labels.Sep, spec)
-		}
-		rg.SlaveIPs[slave.ID] = address
-	}
+	rg.frameworkRecords(sj, domain, spec)
+	rg.slaveRecords(sj, domain, spec)
+	rg.listenerRecord(listener, ns)
+	rg.masterRecord(domain, masters, sj.Leader)
 
 	// complete crap - refactor me
 	for _, f := range sj.Frameworks {
 		fname := labels.DomainFrag(f.Name, labels.Sep, spec)
-		if address, ok := hostToIP4(f.PID.Host); ok {
-			a := fname + "." + domain + "."
-			rg.insertRR(a, address, "A")
-			srv := net.JoinHostPort(a, f.PID.Port)
-			rg.insertRR("_framework._tcp."+a, srv, "SRV")
-		}
 
+		// insert taks records
+		tail := fname + "." + domain + "."
 		for _, task := range f.Tasks {
 			hostIP, ok := rg.SlaveIPs[task.SlaveID]
+
 			// skip not running or not discoverable tasks
 			if !ok || (task.State != "TASK_RUNNING") {
 				continue
 			}
 
-			tname := spec(task.Name)
-			sid := slaveIDTail(task.SlaveID)
-			tag := hashString(task.ID)
-			tail := fname + "." + domain + "."
+			// context used to build domain names
+			context := struct {
+				TaskName string
+				TaskID   string
+				SlaveID  string
+			}{
+				SlaveID:  slaveIDTail(task.SlaveID),
+				TaskID:   hashString(task.ID),
+				TaskName: spec(task.Name),
+			}
 
-			// A records for task and task-sid
-			arec := tname + "." + tail
+			// insert canonical A records
+			trec := context.TaskName + "-" + context.TaskID + "-" + context.SlaveID + "." + tail
```

Comment:
Eventually I'd like to have these name constructions extracted into well tested, single purpose functions.